gotooltest/setup.go

```diff
@@ -7,25 +7,70 @@
 package gotooltest
 
 import (
+	"bytes"
 	"fmt"
 	"go/build"
 	"os/exec"
 	"path/filepath"
 	"regexp"
 	"runtime"
 	"strings"
+	"sync"
 
 	"github.com/rogpeppe/go-internal/imports"
 	"github.com/rogpeppe/go-internal/testscript"
 )
 
 var (
 	goVersionRegex = regexp.MustCompile(`^go([1-9][0-9]*)\.(0|[1-9][0-9]*)$`)
+
+	goEnv struct {
+		goroot      string
+		gocache     string
+		goversion   string
+		releaseTags []string
+		once        sync.Once
+		err         error
+	}
 )
 
-type testContext struct {
-	goroot  string
-	gocache string
+func initGoEnv() {
+	var err error
+
+	var stderr bytes.Buffer
+	cmd := exec.Command("go", "list", "-f={{context.ReleaseTags}}", "runtime")
+	cmd.Stderr = &stderr
+	out, err := cmd.Output()
+	if err != nil {
+		goEnv.err = fmt.Errorf("failed to determine release tags from go command: %v\n%s", err, stderr.String())
+		return
+	}
+	goEnv.releaseTags = strings.Split(strings.Trim(strings.TrimSpace(string(out)), "[]"), " ")
+
+	getGoEnv := func(name string) (string, error) {
+		out, err := exec.Command("go", "env", name).CombinedOutput()
+		if err != nil {
+			return "", fmt.Errorf("go env %s: %v (%s)", name, err, out)
+		}
+		return strings.TrimSpace(string(out)), nil
+	}
+
+	goEnv.goroot, err = getGoEnv("GOROOT")
+	if err != nil {
+		goEnv.err = err
+		return
+	}
+	goEnv.gocache, err = getGoEnv("GOCACHE")
+	if err != nil {
+		goEnv.err = err
+		return
+	}
+
+	version := goEnv.releaseTags[len(goEnv.releaseTags)-1]
+	if !goVersionRegex.MatchString(version) {
+		goEnv.err = fmt.Errorf("invalid go version %q", version)
```

Comment:
You should probably return after this.
It's quite easy to make this mistake, so perhaps change initGoEnv to return an error and do:

    goEnv.once.Do(func() {
        goEnv.err = initGoEnv()
    })

?

Reply:
Thanks, that's a nicer pattern.